# K Nearest Neighbors Project

In this project I'll work with the famous iris dataset from scikit-learn and use knn model to predict wether a given flower is Setosa, Versicolour or Virginica.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [25]:
iris_data = pd.read_csv('iris.data')
iris_data.columns = ['Sepal Length', 'Sepal Width', 'Petal Length', 'Petal Width', 'Class']

In [26]:
iris_data.head()

Sepal Length  Sepal Width  Petal Length  Petal Width        Class
0           4.9          3.0           1.4          0.2  Iris-setosa
1           4.7          3.2           1.3          0.2  Iris-setosa
2           4.6          3.1           1.5          0.2  Iris-setosa
3           5.0          3.6           1.4          0.2  Iris-setosa
4           5.4          3.9           1.7          0.4  Iris-setosa

This dataset is already filtered so it looks like we have no work in the data analisys part and we can move on to training and testing our model.

In [27]:
from sklearn.model_selection import train_test_split

In [30]:
X = iris_data.drop('Class', axis=1)
y = iris_data['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [31]:
from sklearn.neighbors import KNeighborsClassifier

In [32]:
knn_model = KNeighborsClassifier(n_neighbors=5)

In [33]:
knn_model.fit(X_train, y_train)

KNeighborsClassifier()

In [34]:
predictions = knn_model.predict(X_test)

In [35]:
from sklearn.metrics import confusion_matrix, classification_report

In [36]:
print(confusion_matrix(y_test, predictions))
print('\n')
print(classification_report(y_test, predictions))

[[15  0  0]
 [ 0 19  0]
 [ 0  2  9]]


                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        15
Iris-versicolor       0.90      1.00      0.95        19
 Iris-virginica       1.00      0.82      0.90        11

       accuracy                           0.96        45
      macro avg       0.97      0.94      0.95        45
   weighted avg       0.96      0.96      0.95        45



We can see that knn was a right choice for this dataset, we got 96% accuracy with perfect prediction of the setosa flower.